api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -812,6 +813,29 @@ message Cluster {
   // of 0 would indicate that none of the timeout was used or that the timeout was infinite. A value
   // of 100 would indicate that the request took the entirety of the timeout given to it.
   bool track_timeout_budgets = 47;
+
+  // Optional customization and configuration of upstream connection pool, and
+  // upstream type.
+  //
+  // Currently this field only applies for HTTP traffic but is designed for
+  // eventual use for custom TCP upstreams.
+  //
+  // For HTTP traffic, Envoy will generally take downstream HTTP and send it
+  // upstream as upstream HTTP, using the connection pool
+  // `envoy.filters.connection_pools.http.http` and the codec from `http2_protocol_options`
+  //
+  // For routes where CONNECT termination is configured, Envoy will take
+  // downstream CONNECT requests and forward the CONNECT payload upstream over
+  // raw TCP, using `envoy.filters.connection_pools.http.tcp`.
+  //
+  // The default pool used in Envoy `envoy.filters.connection_pool.http.default`, which
+  // creates the HTTP upstream for most HTTP requests, and the TCP upstream if
+  // CONNECT termination is configured.
+  //
+  // If users desire custom connection pool or upstream behavior, for example terminating
+  // CONNECT only if a custom filter indicates it is appropriate, the custom factories
+  // can be registered and configured here.
```

Comment:
Are there some example configs/extensions that can be plugged in here today?

All extensions must have a unique type in the API, providing a type URL to identify them. Using names like `envoy.filters.connection_pools.http.http` is now deprecated and redundant with type URL for typed extensions.

Reply:
we're doing this for TCP and HTTP upstreams.  To enable things like https://github.com/envoyproxy/envoy/pull/11112/files